File: 0701-0800/0718-Maximum Length of Repeated Subarray/0718-Maximum Length of Repeated Subarray.cpp
class Solution {
public:
    int findLength(vector<int>& A, vector<int>& B) {
        int m = A.size();
        int n = B.size();
        vector<vector<int>> dp(1 + m, vector<int>(1 + n));
        int maxLen = 0;
        for (int i = 1; i <= m; ++i) {
            for (int j = 1; j <= n; ++j) {
                if (A[i - 1] == B[j - 1]) {
                    dp[i][j] = 1 + dp[i - 1][j - 1];
                    maxLen = max(maxLen, dp[i][j]);
                }               
            }
        }
        
        return maxLen;
    }
};

class Solution {
public:
    int findLength(vector<int>& A, vector<int>& B) {
        int m = A.size();
        int n = B.size();
        vector<int> dp(1 + n);
        int maxLen = 0;
        for (int i = 1; i <= m; ++i) {
            for (int j = n; j >= 1; --j) {
                if (A[i - 1] == B[j - 1]) {
                    dp[j] = 1 + dp[j - 1];
                    maxLen = max(maxLen, dp[j]);
                }
                else {
                    dp[j] = 0;
                }                
            }
        }
        
        return maxLen;
    }
};

File: 0701-0800/0718-Maximum Length of Repeated Subarray/0718-Maximum Length of Repeated Subarray.py
class Solution:
    def findLength(self, A: List[int], B: List[int]) -> int:
        dp = [0] * (1 + len(B))
        maxLen = 0
        for i in range(1, 1 + len(A)):
            for j in range(len(B), 0, -1):
                if A[i - 1] == B[j - 1]:
                    dp[j] = 1 + dp[j - 1]
                    maxLen = max(maxLen, dp[j])
                else:
                    dp[j] = 0
        return maxLen
